[Alerting][Docs] Adds Alerting & Task Manager Scalability Guidance & Health Monitoring (#91171)
elastic/kibana: docs/user/production-considerations/task-manager-health-monitoring.asciidoc

[role="xpack"]
[[task-manager-health-monitoring]]
== Task Manager health monitoring

++++
<titleabbrev>health monitoring</titleabbrev>
++++

> The {kib} Task Manager has an internal monitoring mechanism keeping track of a variety of metrics which can be consumed via both a health monitoring API, and the {kib} Server Log.

Author: Anyone who needs to monitor the health, so presumably engineers deploying Kibana who are in charge of administrating the system.

> The health monitoring API is designed to be fast, providing a reliable endpoint which can be monitored.

Author: The point we're trying to convey here is aimed at individuals in charge of ensuring Kibana and Alerting are up and running.
As an engineer in charge of DevOps I would want to know that the endpoint I'm using to monitor the system was designed for that usage and is reliable and fast enough to fill that role.

> Consuming this endpoint doesn't cause additional load, but rather it returns the latest health checks made by the system. This design enables consumption by external services at a regular cadance without the concern of additional load to the system.

Author: That's up to the customer 🤷 
Monitoring endpoints are designed to be consumed in order to verify the system is up and running.

Many of companies will be using Kibana and Alerting to monitor their services, but they won't want to rely on Kibana/Alerting to monitor itself (if it's down it can't tell you it's down), so they'll use another service, such as Nagios, to consume this endpoint and alert them when Kibana/Alerting is down.

> Consuming this endpoint doesn't cause additional load, but rather it returns the latest health checks made by the system. This design enables consumption by external services at a regular cadance without the concern of additional load to the system.

Author: I've added the word "monitoring", so it's "external monitoring services"

Each {kib} instance exposes its own endpoint at:

[source,sh]
--------------------------------------------------
$ curl -X GET api/task_manager/_health
--------------------------------------------------
// KIBANA

Monitoring the `_health` endpoint of each {kib} instance in the cluster is the recommended method of ensuring confidence in mission critical services such as Alerting and Actions.

[float]
[[task-manager-configuring-health-monitoring]]
=== Configuring the Monitored Health Statistics

The health monitoring api monitors Task Manager's performance out of the box, but certain performance considerination are deployment specific and are configurable as a result.

> Health thresholds are configurable via the <<task-manager-health-settings,`xpack.task_manager.monitored_task_execution_thresholds`>> setting.

Author: Sounds great 👍

This configuration specifies the threshold of failed task executions which, once exceeded, sets a health status of `warn` or `error` on the corresponding task type execution status.
This setting can be set as either the default level, which is applied to all task types in the system, or at a custom task type specific level. 

By default, this setting is configured to mark the health of every task type as `warning` when it exceeds 80% failed executions, and as `error` at 90%.
This value can be set to any number between 0 to 100, and a threshold is hit when the value *exceeds* this number.

> This means that you can avoid setting the status to `error` by setting the threshold at 100, or hit `error` the moment any task fails by setting the threshold to 0 (as it will exceed 0 once a single failure occurs).

Reviewer: ```suggestion
To avoid a status of `error`, set the threshold at 100.  To hit `error` the moment any task fails, set the threshold to 0.
```

> Custom configurations allow you to set lower thresholds for task types you might consider critical, such as alerting tasks, which you might wish to detect sooner in an external monitoring service.

Reviewer: ```suggestion
Create a custom configuration to set lower thresholds for task types you consider critical, such as alerting tasks that you want to detect sooner in an external monitoring service.
```

[source,yml]
----
xpack.task_manager.monitored_task_execution_thresholds:
  default: # <1>
    error_threshold: 70
    warn_threshold: 50
  custom:
    "alerting:.index-threshold": # <2>
      error_threshold: 50
      warn_threshold: 0
----

> <1> default configuration, setting a system wide `warn` threshold at 50%, and `error` at 70% failure rate

Reviewer: ```suggestion
<1> A default configuration that sets the system-wide `warn` threshold at a 50% failure rate, and `error` at 70% failure rate.
```

> <2> custom configuration for the `alerting:.index-threshold` task type, setting a system wide `warn` threshold at 0% (which will set a `warn` status the moment any task of that type fails), and `error` at 50% failure rate

Reviewer: ```suggestion
<2> A custom configuration for the `alerting:.index-threshold` task type that sets a system wide `warn` threshold at 0% (which sets a `warn` status the moment any task of that type fails), and `error` at a 50% failure rate.
```

[float]
[[task-manager-consuming-health-stats]]

> === Consuming Health Stats

Reviewer: ```suggestion
=== Consuming health stats
```

> The `health` api is best consumed by via the `/api/task_manager/_health` endpoint.

Reviewer: ```suggestion
The health API is best consumed by via the `/api/task_manager/_health` endpoint.
```

> Additionally, the metrics are logged out into the {kib} `DEBUG` logger at a regular cadence.

Reviewer: ```suggestion
Additionally, the metrics are logged in the {kib} `DEBUG` logger at a regular cadence.
```

Reviewer: Is it necessary to have DEBUG in all caps?

Author: Haha not at all, it's just appears that way in the log, which is why I wrote it that way,

> If you wish to enable Task Manager DEBUG logging in your {kib} instance, you will need to add the following to your `Kibana.yml`:

Reviewer: ```suggestion
To enable Task Manager DEBUG logging in your {kib} instance, add the following to your `kibana.yml`:
```

> ```

Reviewer: This code block should be formatted to have a black background.

logging:
  loggers:
      - context: plugins.taskManager
        appenders: [console]
        level: debug
```

> Please bear in mind that these stats are logged as often as your <<task-manager-settings,`xpack.task_manager.poll_interval`>> setting, which means it could add substantial noise to your logs.

Reviewer: ```suggestion
These stats are logged based the number of milliseconds set in your <<task-manager-settings,`xpack.task_manager.poll_interval`>> setting, which means it could add substantial noise to your logs.
```

> We would recommend only enabling this level of logging temporarily.

Reviewer: ```suggestion
Only enable this level of logging temporarily.
```

Reviewer: Recommend combining this sentence with the previous sentence for a single paragraph.

[float]
[[making-sense-of-task-manager-health-stats]]

> === Making Sense of Task Manager Health Stats

Reviewer: ```suggestion
=== Making sense of Task Manager health stats
```

The health monitoring API exposes three sections: `configuration`, `workload` and `runtime`:

[cols="2"]
|===

a| Configuration

> | This section summarizes Task Manager's current configuration, including dynamic configurations which change over time, such as `poll_interval` and `max_workers` which can adjust in reaction to changing load on the system.

Reviewer: ```suggestion
| This section summarizes the current configuration of Task Manager.  This includes dynamic configurations that change over time, such as `poll_interval` and `max_workers`, which can adjust in reaction to changing load on the system.
```

a| Workload

> | This section summarizes the work load across the cluster, listing the tasks in the system, their types and what their current status is.

Reviewer: ```suggestion
| This section summarizes the work load across the cluster, including the tasks in the system, their types, and current status.
```

a| Runtime

> | This section tracks excution performance of Task Manager, tracking task _drift_, worker _load_, and execution stats broken down by type, including duration and execution results

Reviewer: ```suggestion
| This section tracks excution performance of Task Manager, tracking task _drift_, worker _load_, and execution stats broken down by type, including duration and execution results.
```

|===

> Each section has a `timestamp` and a `status` which indicates when the last update to this setion took place and whether the health of this section was evaluated as `OK`, `Warning` or `Error`.

Reviewer: ```suggestion
Each section has a `timestamp` and a `status` that indicates when the last update to this section took place and whether the health of this section was evaluated as `OK`, `Warning` or `Error`.
```

> The root has its own `status` which indicate the `status` of the system overall.

Reviewer: ```suggestion
The root `status` indicates the `status` of the system overall.
```

> By monitoring the `status` of the system overall, and the `status` of specific task types of interest, it is easy to evaluate the health of the {kib} Task Management system.

Reviewer: ```suggestion
By monitoring the `status` of the system overall, and the `status` of specific task types of interest, you can evaluate the health of the {kib} Task Management system.
```
